### PR TITLE
Adding automatic anisotropic downsampling option to conversion

### PR DESCRIPTION
**Pull request by MartinBuessemeyer**
This PR adds a Method to `downsampling.py` that creates multiple anisotropic downsampled Magnifications in one call and adds a `argparse` option to the main script in order to enable wkw-conversion and anisotropic downsampling in one step.

Needs to be tested.

### CODE REVIEW

**Comment by MartinBuessemeyer** [comment]
**Added:**
-  Testscript `simple_anisotropic_tiff_cubing.sh`
- `Mag` `<=` implementation
- adaptive dimension choosing for anisotropic downsampling
- removed a bug in `compress.py`

Ready to be merged I guess.

File: wkcuber/utils.py [hunk 1]
```diff
@@ -61,6 +61,18 @@ def add_verbose_flag(parser):
     parser.set_defaults(verbose=True)
 
 
+def add_anisotropic_flag(parser):
+    parser.add_argument(
+        "--anisotropic",
+        "-aniso",
```
Comment (normanrz) [inline wkcuber/utils.py]: The single dash shorthands, should only be single chars, i.e. `-a`

File: wkcuber/__main__.py [hunk 2]
```diff
@@ -89,15 +90,31 @@ def create_parser():
...
-        args.target_path,
-        args.layer_name,
-        Mag(1),
-        Mag(args.max_mag),
-        "default",
-        DEFAULT_EDGE_LEN,
-        not args.no_compress,
-    )
+    if args.anisotropic:
+        scale = tuple(float(x) for x in args.scale.split(","))
```
[the remark this answers is not in the document]
Comment (MartinBuessemeyer) [reply wkcuber/__main__.py]: True. I pulled the calculation in front of the if-statements. :+1:

File: wkcuber/utils.py [hunk 3]
```diff
@@ -61,6 +61,18 @@ def add_verbose_flag(parser):
...
+        help="Anisotropic downsampling",
```
[the remark this answers is not in the document]
Comment (MartinBuessemeyer) [reply wkcuber/utils.py]: That's right. I forgot about that.

File: wkcuber/downsampling.py [hunk 4]
```diff
@@ -455,6 +456,61 @@ def downsample_mags(
...
+        )
+        prev_mag = target_mag
+        target_mag = get_next_anisotropic_mag(target_mag, scale)
+
+
+def get_next_anisotropic_mag(mag, scale):
+    max_index, min_index = detect_larger_and_smaller_dimension_by_size(scale)
+    scale_increase = [1, 1, 1]
+    if (
+        mag.to_array()[min_index] * 2 * scale[min_index]
```
[the remark this answers is not in the document]
Comment (MartinBuessemeyer) [reply wkcuber/downsampling.py]: Yes indeed :D

File: wkcuber/downsampling.py [hunk 5]
```diff
@@ -455,6 +456,61 @@ def downsample_mags(
...
+    return Mag(
+        [
+            mag.to_array()[0] * scale_increase[0],
+            mag.to_array()[1] * scale_increase[1],
+            mag.to_array()[2] * scale_increase[2],
+        ]
+    )
+
+
+def detect_larger_and_smaller_dimension_by_size(scale):
```
[the remark this answers is not in the document]
Comment (MartinBuessemeyer) [reply wkcuber/downsampling.py]: Yea. The name is quite long.

**Comment by philippotto** [comment]
Adding a unit test for the "get next magnification" code would also be cool. The test could simply calculate the target magnifications for some specific resolutions (e.g., `11,11,24`, `5,5,5`, `20,10,20`, `10,10,11` etc.) without doing the actual downsampling. That way, we cover more edge cases.

**Comment by normanrz** [comment]
I agree that a more sophisticated mag-determining-algo would be better. However, I would like to postpone that to another PR. I'll need this PR merged by Thursday.

**Comment by philippotto** [comment]
Thanks for adding the unit tests, @MartinBuessemeyer :+1:

**Comment by normanrz** [comment]
A code style remark: The scales (e.g. `(10.5, 10.5, 24)`) and mag definitions (e.g. `Mag((2,2,1))`) should be represented through tuples `(..)` instead of lists `[..]`

File: tests/test_downsampling.py [hunk 6]
```diff
@@ -181,3 +183,17 @@ def test_downsample_multi_channel():
     assert np.any(target_buffer != 0)
 
     assert np.all(target_buffer == joined_buffer)
+
+def test_anisotropic_mag_calculation():
+    scale_test = [(10.5, 10.5, 24), (10.5, 10.5, 21), (10.5, 24, 10.5), (24, 10.5, 10.5), (10.5, 10.5, 10.5)] * 2
+    mags_test = [Mag(1), Mag(1), Mag(1), Mag(1), Mag(1),
+                 Mag([2, 2, 1]), Mag([2, 2, 1]), Mag([2, 1, 2]), Mag([1, 2, 2]), Mag(2)]
+    mags_expected = [Mag([2, 2, 1]), Mag([2, 2, 1]), Mag([2, 1, 2]), Mag([1, 2, 2]), Mag([2, 2, 2]),
```
Comment (philippotto) [inline tests/test_downsampling.py]: The test cases would be a bit easier to read if there were structured like that:
```
[
  [input_mag1, expected_mag1],
  [input_mag2, expected_mag2],
  ...
]
```

**Comment by MartinBuessemeyer** [comment]
Merge?

**Comment by philippotto** [comment]
> Merge?

Yes, feel free to merge when the CI build has passed. Please use the "squash and merge" option so that the commit history on master stays concise :)
